### ISSUE TITLE
Unable to upgrade a single package unless --no-best is used

### ISSUE THREAD

**Issue by Lamieur** [rpm-software-management/dnf5#1106]
F38, dnf5-5.1.7-1.fc38.x86_64

When running `dnf5 upgrade <package>`, dnf5 wants to upgrade any dependency of any package being upgraded.

The same happens with other methods of limiting the upgrade's scope, like `dnf5 upgrade --security` additionally pulls all dependencies of all packages with security updates.

Old dnf 4 works as expected:
```
# LANG=C dnf upgrade discord
(...)
Upgrading:
 discord        x86_64        0.0.38-1.fc38        rpmfusion-nonfree-updates         70 M

Transaction Summary
==========================================================================================
Upgrade  1 Package
```
dnf5 wants to do much more:
```
# dnf5 upgrade discord
(...)
Upgrading:                                                                                
 discord                        x86_64 0.0.38-1.fc38              rpmfusion-nonf 219.5 MiB
  replacing discord             x86_64 0.0.37-1.fc38              rpmfusion-nonf 219.5 MiB
 glibc                          i686   2.37-16.fc38               updates          5.5 MiB
  replacing glibc               i686   2.37-14.fc38               updates          5.5 MiB
(...)
Transaction Summary:
 Upgrading:        15 packages
 Replacing:        15 packages
```
This is not a full upgrade (that would upgrade 93 packages + kernel stuff) nor an upgrade --security (that would upgrade 59 + kernel - hitting the exact same issue, as these packages do NOT offer security fixes - dnf upgrade --security offers just 35 + kernel).

Looks like dnf upgrade <package> additionally pulls updates for the package's dependencies, even if they would be satisfied without any more updates.

And in fact, I can make it work like dnf4:
```
# dnf5 update discord --no-best
Updating and loading repositories:
Repositories loaded.
Package                    Arch   Version                    Repository               Size
Upgrading:                                                                                
 discord                   x86_64 0.0.38-1.fc38              rpmfusion-nonfree-u 219.5 MiB
   replacing discord       x86_64 0.0.37-1.fc38              rpmfusion-nonfree-u 219.5 MiB

Transaction Summary:
 Upgrading:         1 packages
 Replacing:         1 packages
```

dnf5 man page doesn't say --best is the default (and its description covers a very different case), and even if it was, same man page specifically says "Note  that the use of the newest available version is only guaranteed for the packages directly requested (e.g. as a command line  arguments),  and the  solver may use older versions of dependencies to meet their requirements."

So why would --no-best fix this issue? Its description says "Do not limit the transaction to the best candidates only.", which sounds right, but in this case, it starts meaning "prevent dnf5 from forcing best candidates of packages unrelated to the requested action" ;) It still picks the best candidate for my single package upgrade, fortunately.

I was googling for an explanation, but couldn't find anything about this difference between dnf4 and 5.

I don't think this is the expected behavior. It's different from yum/dnf4, at first glance undocumented, and not what users expect.

But if, for some reason, it is expected, perhaps it can be documented more clearly, so googling for "dnf5 upgrade single package" finds anything :)

PS. `dnf5 upgrade --minimal discord`, and even adding all the packages that it wanted to update along discord, `dnf5 upgrade --minimal discord glibc\* libnsl` says "Nothing to do." - because (as previously mentioned), none of them are marked as bugfix/enhancement/security. Took me a while to figure --no-best (which isn't mentioned in dnf5-upgrade(8) at all) :)

**Comment by kontura** [comment]
I believe this is a consequence of https://github.com/rpm-software-management/dnf5/pull/779 and is expected but you are correct: it is different from dnf4 and it is not documented (that is my mistake and we should fix that).

Basically the change is that dnf5 prefers latest dependency versions over smaller transactions.

For more context the issue we are trying to solve is installing non-latest dependencies for an `install` command. We had several reports for it like https://bugzilla.redhat.com/show_bug.cgi?id=2150102.

We have decided to try changing the behavior uniformly across all commands.

If I understand correctly the problem for you is mostly that it is not documented correctly but I am curious what do you think about the new behavior? Does it cause issues for your workflows or is it just unexpected/different?

**Comment by Lamieur** [comment]
In this case, I simply wanted to update a single package (because it stops working as a client whenever another update is released by the vendor) while keeping everything not updated, because I want to have a few days' backlog for an attempt of dnf5 --releasever=39 upgrade :) There is a command for that, it didn't work as advertised, I even found a work-around within dnf5.

But I'm thinking about this getting into RHEL10+ and what it would mean in a more "serious" environment. I work as a RHEL admin at a place where it's not unusual to install or update different packages at different schedules. We already have local snapshots of RHEL updates, so there's a baseline we install in a testing environment which has to pass tests before installing the exact same updates in production (and nothing newer, aka untested). But we also have multiple repos of different kinds of software, maintained by a multitude of teams, which don't always follow this baseline snapshot paradigm. And they get tested and rolled out independently of each other or OS/security updates.

This means that it's still possible that there is an update available for a package that wasn't tested and signed off (therefore I'm not allowed to install it yet), but would be picked up by dnf install/update of something else.

This leads us to using dnf --disablerepo=\* --enablerepo=X install Y, but even that won't be enough, in the future I'll need --no-best as well (and carefully check other side-effects of this switch).

Personally I think yum-dnf4 way of "if dependencies are met, don't do more than the user asks for; but if I need to install a dependency, use the newest candidate" is intuitive and pretty transparent.

dnf5's new thinking of "I'm installing X, why not update all of its dependencies (even if it doesn't ask for specific version) as well?" is not intuitive - it's impossible for me as a user to predict what will happen if I want to install/update a specific package. Not a big deal for me as a Fedora user (I'll update all of that soon anyways), but still counter-intuitive. I won't have to deal with it at work for years, so right now it's just a curiosity.

So, in short, even if it's impossible to fix without breaking dependency pulling / destroying performance, it's an inconsistency from previous dnf, so it needs better documentation. If someone really needs to install/update a single package and dnf4 can just do it™, there needs to be easily "googlable" solution for people switching to dnf5.

Otherwise users will soon be bombarded with guides on "how to install X" which start with "dnf5 doesn't let us do it, so let's curl | bash, or rpm -U http://..." ;) It's already the biggest problem for Linux as a whole - all the bad instructions popping up whenever you search for anything...

**Comment by voxik** [comment]
I was bitten by this and I must say this is pretty annoying behavior. One could just wave it off when upgrading package, but it is much bigger problem when downgrading package. Recently, I wanted to downgrade `mpv`, because it have not worked, while the transaction offered me:

~~~
 Installing:       11 packages
 Upgrading:        50 packages
 Replacing:        52 packages
 Downgrading:       1 packages
~~~

After all, I was able to downgrade just the `mpv`, but DNF have not made this easy.

Maybe the #779 should be the case just for installation. Or this should be configurable and DNF should even suggest the option to override this behavior.

**Comment by Lamieur** [comment]
Hehe, remember when I mentioned how the reason for me not willing to update more than the single package was because I was preparing for F39 upgrade with dnf5?

Guess what! I had to use --no-best for `dnf5 --releasever=39 upgrade` as well, otherwise it would refuse to work at all (dnf 4 would simply reinstall a single package that dnf5 wouldn't tell me about, but failed on updating dependencies). Most curious of all - after that operation, neither of them wanted to do anything else. It just worked with --no-best and I don't know how. I don't know if the package has been reinstalled, but the program works fine and both dnfs agree. Has it been reinstalled? I can't say - I wasn't paying enough attention - I assumed I will need dnf4 upgrade afterwards, but no, dnf5 --no-best made everything happy, I'm typing this from an upgraded system untouched by dnf other than dnf5.

Probably unrelated to a single package upgrade, but confirming that, currently, anyone should just alias dnf5="dnf5 --no-best" becauses it fixes all the observable issues ;)

Again: you do NOT want this kind of message to be posted on the stackoverflows of the world. I may joke about this around here, but I'm really concerned about the fallout of such an undocumented (and unexpected) behavior.

**Comment by voxik** [comment]
Again struggling with this. This time trying to install `rubygem-json`:

~~~
$ sudo dnf install rubygem-json --no-best
Repositories loaded.
Package                         Arch     Version                       Repository          Size
Upgrading:                                                                                     
 bash                           x86_64   5.2.21-2.fc40                 rawhide          8.2 MiB
  replacing bash                x86_64   5.2.15-5.fc39                 rawhide          8.2 MiB
 glibc                          x86_64   2.38.9000-29.fc40             rawhide          6.7 MiB
  replacing glibc               x86_64   2.38.9000-16.fc40             rawhide          6.6 MiB
 glibc-common                   x86_64   2.38.9000-29.fc40             rawhide          1.0 MiB
  replacing glibc-common        x86_64   2.38.9000-16.fc40             rawhide          1.0 MiB
 glibc-gconv-extra              x86_64   2.38.9000-29.fc40             rawhide          7.8 MiB
  replacing glibc-gconv-extra   x86_64   2.38.9000-16.fc40             rawhide          7.8 MiB
 glibc-langpack-cs              x86_64   2.38.9000-29.fc40             rawhide          2.9 MiB
  replacing glibc-langpack-cs   x86_64   2.38.9000-16.fc40             rawhide          2.9 MiB
 glibc-langpack-en              x86_64   2.38.9000-29.fc40             rawhide          5.7 MiB
  replacing glibc-langpack-en   x86_64   2.38.9000-16.fc40             rawhide          5.7 MiB
 libgcc                         x86_64   13.2.1-6.fc40                 rawhide        230.3 KiB
  replacing libgcc              x86_64   13.2.1-4.fc40                 rawhide        230.3 KiB
 libxcrypt                      x86_64   4.4.36-3.fc40                 rawhide        266.8 KiB
  replacing libxcrypt           x86_64   4.4.36-2.fc39                 rawhide        266.8 KiB
 ruby                           x86_64   3.2.2-183.fc40                rawhide         91.9 KiB
  replacing ruby                x86_64   3.2.2-182.fc40                rawhide         91.9 KiB
 ruby-default-gems              noarch   3.2.2-183.fc40                rawhide         74.4 KiB
  replacing ruby-default-gems   noarch   3.2.2-182.fc40                rawhide         74.4 KiB
 ruby-libs                      x86_64   3.2.2-183.fc40                rawhide         14.0 MiB
  replacing ruby-libs           x86_64   3.2.2-182.fc40                rawhide         14.0 MiB
 rubygem-io-console             x86_64   0.6.0-183.fc40                rawhide         37.1 KiB
  replacing rubygem-io-console  x86_64   0.6.0-182.fc40                rawhide         37.1 KiB
 rubygem-psych                  x86_64   5.0.1-183.fc40                rawhide        139.9 KiB
  replacing rubygem-psych       x86_64   5.0.1-182.fc40                rawhide        139.9 KiB
 rubygems                       noarch   3.4.10-183.fc40               rawhide        967.6 KiB
   replacing rubygems           noarch   3.4.10-182.fc40               rawhide        967.5 KiB
Installing:                                                                                    
 rubygem-json                   x86_64   2.6.3-204.fc39                rawhide        164.3 KiB
Installing dependencies:                                                                       
 zlib-ng-compat                 x86_64   2.1.5-1.fc40                  rawhide        137.8 KiB
   replacing zlib               x86_64   1.2.13-5.fc40                 rawhide        206.6 KiB

Transaction Summary:
 Installing:        2 packages
 Upgrading:        14 packages
 Replacing:        15 packages

Total size of inbound packages is 12 MiB. Need to download 12 MiB.
After this operation 108 KiB will be used (install 48 MiB, remove 48 MiB).
Is this ok [y/N]: 
~~~

Yes, I know that my system was not updated in a while, but why should I update those packages? It does not make a sense to update `glibc` which might make more troubles then installing `rubygem-json` which would by a chance was not somehow compatible with my Ruby.

But what is even more annoying, the `--no-best` flag does not seems to change this behavior. So how to really install just `rubygem-json`? Do I really need to use something like `dnf install rubygem-json -x rubygems -x glibc\* -x \*lib\*`?

**Comment by voxik** [comment]
Is there a way to workaround it? `--no-best` is not of any help actually :/

**Comment by j-mracek** [comment]
I am proposing to make `libsolv_solver.set_flag(SOLVER_FLAG_FOCUS_BEST, 1);` conditional.

Create `boolean` configuration option for FOCUS_BEST with default `false` in upstream.

The current behavior is problematic, because it force to upgrade whole dependency tree. In combination with `dnf5 upgrade --minimal` it is 100% unwanted.

I also suggest to fix it as soon as possible, because it creates more negative UX then positive.

@jan-kolarik May I ask you to add the issue to the current plan?
